### ISSUE TITLE
napari RGB images becomes 3D stack in imagej with x = color component

### ISSUE THREAD

**Issue by NicoKiaru** [imagej/pyimagej#221]
An RGB image type in napari becomes a XYZ image in ImageJ with X = color component, Y = Y, Z = X.

How to reproduce:
- open a sample RGB from napari (File > Open Sample > napari > Astronaut (RGB)),
- send it to imagej - it will result in a very elongated image:

![image](https://user-images.githubusercontent.com/20223054/188155043-32995e17-c98a-4150-91ad-c694b2310998.png)

**Comment by ctrueden** [comment]
@gselzer @NicoKiaru Unfortunately, there is no axis metadata that goes along with scikit-image sample images. They are just plain `numpy.ndarray` objects. PyImageJ's convention when converting a plain numpy array into Java/ImgLib2 is simply to reverse the order of the dimensions. The astronaut's shape is (512, 512, 3), which inverts to (3, 512, 512) `ImgView` i.e. `Img` i.e. still no axis metadata. Then when you throw an `Img` at `ij.ui().show(...)`, it treats the first two dimensions as X and Y.

One solution is to wrap the numpy array into an xarray with proper dimensional labels before converting:
```python
>>> xastro = xarray.DataArray(astronaut(), name='xastronaut', dims=('row', 'col', 'ch'))
>>> xastro.shape
(512, 512, 3)
>>> xastro.dims
('row', 'col', 'ch')
>>> jastro = ij.py.to_java(xastro)
>>> [jastro.dimension(d) for d in range(jastro.numDimensions())]
[512, 512, 3]
>>> [jastro.axis(d).type().getLabel() for d in range(jastro.numDimensions())]
['X', 'Y', 'Channel']
```

See also section [6.7 Passing image data from Python to Java](https://pyimagej.readthedocs.io/en/latest/6-Working-with-Images.html#passing-image-data-from-python-to-java) of the PyImageJ documentation.

**Comment by imagejan** [comment]
It seems to be a convention in Python land to treat an array as RGB image **iff** the third dimension has a size of 3 (i.e. `(m, n, 3)` gets interpreted as `XYC` where as `(m, n, 10)` would be `ZYX`). Is this observation correct? And if yes, should the special-casing be copied in PyImageJ?

**Comment by ctrueden** [comment]
> Is this observation correct?

I believe so. But this is the sort of "magic" that I believe @jni frowns upon (right @jni?).

> And if yes, should the special-casing be copied in PyImageJ?

I am reluctant. In microscopy it is rare for the X dimension to be length 3, yes, but I vaguely recall seeing an issue where someone did have this in a scientific image, and ran into problems with scikit-image and/or napari because of it. But ultimately I would defer to @jni on it: we can match this convention if the scientific Python experts deem it to be best match the principle of least astonishment.

**Comment by jni** [comment]
I would say *don't* match it. tbh we have found that these assumptions are the source of many small bugs (like the one reported here). It remains a goal of the scientific Python imaging community to add axis annotations, and recent efforts by the OME folks are starting to bear fruit there. So, I think it's best to wait until we have proper annotations, than to perpetuate a poorly documented and sometimes problematic convention.

**Comment by elevans** [comment]
I also agree with @jni in that we should **not**  create special case/match arrays with a third dimension of size `3`. As I see it we should *do nothing*. There is no metadata here to make any assumptions, therefore we should not make any. That being said, I think we can mitigate the effects of these kind of dimension mangling bugs by adding a convenience method to reorder and manipulate the image and it's axis.

I talked to @gselzer and we circled around two ways to do this:

1. Add an optional `order` parameter to the scyjava converters.
2. Add a convenience methods for permuting/modifying the image axes.

I prefer the second option as we already do `RandomAccessibleInterval` permuting when converting RAIs to `xarray.DataArray` objects:

https://github.com/imagej/pyimagej/blob/7d94c726de8bd3abe6417e8e75665c221d125c24/src/imagej/__init__.py#L740-L768

The `dims.py` module should be easy to modify to accept user defined axes for the permute operation.

**Comment by elevans** [comment]
## Update

I pushed two new branches today to [`pyimagej`](https://github.com/imagej/pyimagej/compare/master...to-xarray) and [`scyjava`](https://github.com/scijava/scyjava/tree/to-xarray) called `to-xarray`. Pull both to try out the new stuff. After trying a few approaches and some chats with the team, I settled on wrapping all NumPy images as `xarray.DataArray` with an optional dimension order argument. 

Here's an example of it's usage:

```python
>>> import imagej
>>> from skimage.data import astronaut
>>> ij = imagej.init()
>>> img = ij.py.to_java(astronaut(), dim_order=["row", "col", "ch"]) # can also use ["y", "x", "c"]
>>> img.shape
>>> (512, 512, 3)
```
Under the hood `to_java()` and `to_img()` now checks for `dim_order` kwargs and sends the NumPy array to `_numpy_to_xarray()` instead of `_numpy_to_img()` (no longer used). The end result is, internally, all NumPy arrays are turned into `xarray.DataArray`s with or without dimension labels. If you do not provide a `dim_order` the NumPy array _is not_ reorganized with the channel dimension last. This is the current behavior (_i.e._ the dimensions are in the wrong place for ImageJ). If `dim_order` is supplied then the dimensions are properly reoganized to ImageJ's preference.

I have also added a public `to_xarray()` method for parity with `to_dataset()`. We want users to use metadata with their image data so we should make this as easy as possible.

This can be kind of confusing, so I made this graphic to illustrate the data flow (and to help me navigate `to_java()`).
- Gray = Input or Output image object
- Blue = Existing methods
- Red = New methods added in `to-xarray`
- Dark blue = Methods slated for removal

![to_java](https://user-images.githubusercontent.com/6422847/190250823-8f2e9010-612e-4291-bb14-4482563a5625.svg)

## TODO

This works but there are still bugs/edge cases to deal with.

- [ ] Write tests.
- [ ] Check if methods accept kwargs (right now it's possible to do `ij.py.to_java(5, dim_order=["x", "y", "c"])` which is illegal as `int` does not accept kwargs!!
- [ ] Add method to "guess" dimensions if non are provided and a guess is requested.

**Comment by ctrueden** [comment]
An update on the `to-xarray` work: there are now corresponding PRs, scijava/scyjava#54 and imagej/pyimagej#231.

That said: this issue was originally about the following astonishing behavior in napari-imagej:

> * open a sample RGB from napari (File > Open Sample > napari > Astronaut (RGB)),
> * send it to imagej - it will result in a very elongated image:

And just having a `dim_order` hint in API is not enough, on its own, to avoid this issue. We would also need to expose it in the napari-imagej GUI as an option when clicking the arrow buttons to transfer images between napari and ImageJ. So consequently I am transferring this issue back to the napari-imagej space. Ping pong ping!
